### PR TITLE
Pairwise string distance comparison

### LINKED ISSUE #1994: [FEAT] Allow fuzzy matches on array-valued columns
### Is your proposal related to a problem?

One typical use of array-valued columns is when one known entity has multiple values for a single attribute, a match on any of which is acceptable. For example one person may have multiple telephone numbers or an address history, or even multiple names (married/unmarried names, aliases, etc). While exploding the records outside of splink and post-processing matches is one alternative to array-valued columns, this affects match probabilities and may create memory challenges. 

Currently the only way to compare array-valued columns is to use the array_intersect functions, which require exact matches among the elements of the array.

However, in many circumstances, these values are prone to error and fuzzy matching may be required to identify matches.

### Describe the solution you'd like

I propose a family of matching functions of the form array_min_sim(x, y, simtype) and array_max_sim(x,y,simtype). Here x and y are arrays, and "simtype" is a string identifying one of the existing fuzzy string comparison metrics (alternatively, a separate array function could be implemented for each metric). 

The semantics would be to return the maximum or minimum similarity (alternatively, distance) between an element of array x and an element of array y.  These max/min similarities could be thresholded to get comparison levels.

I'm not sure what the best implementation of this would be. Duckdb may allow an implementation using an iterated call to list_reduce. For example, for array_min_sim, the underlying code could be a list_reduce applied to [Infinity,x] (Infinity prepended to x), with reduce function (a, b) -> min(a, list_reduce( [Infinity,y], (c,d) -> min(c, dist(b,d ) ) .

Alternatively, the arrays could be exploded locally and a full cartesian product could be constructed temporarily (with no other variables to reduce memory overhead), the string comparison run on on the expanded data, then aggregated with min/max; I believe this is similar to what is/was being considered for array valued blocking keys (#1448) (though I can't say I understand the internals enough to understand if this will work!)

### Describe alternatives you've considered

Un-nesting the data before running splink and post-processing is one option, but this affects m-values and creates memory challenges via duplicated records. 

Using array intersections as another alternative, but it demands exact matching.

### PR DESCRIPTION
**Pull request by zmbc**
### Type of PR

- [ ] BUG
- [x] FEAT
- [ ] MAINT
- [ ] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?

This is a follow up to #2195, addressing the PR comments there. Closes #1994.

### Give a brief description for the solution you have provided

As discussed in the prior PR, this mostly models `PairwiseStringDistanceFunctionAtThresholds` and `PairwiseStringDistanceFunctionLevel` off of `DistanceFunctionAtThresholds` and `DistanceFunctionLevel`
respectively.
The main difference is that it is pairwise on an array column (duh) and that it only accepts a small list
of _string_ distance functions and transpiles them, instead of the user passing an arbitrary SQL function.

### PR Checklist

- [x] Added documentation for changes
- [ ] Added feature to example notebooks or tutorial (if appropriate)
- [x] Added tests (if appropriate)
- [x] Updated CHANGELOG.md (if appropriate)
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint_and_format.html)
- [ ] Run the [spellchecker](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/contributing_to_docs.html?h=spellch#spellchecking-docs) (if appropriate)

### CODE REVIEW

**Comment by zmbc** [comment]
Closes #1994

**Review by RobinL — COMMENTED** [review]
Thank - I think this looks good.  Minor comment below about the default argument - bot other than that i think this is good to merge

File: splink/comparison_library.py [hunk 1]
```diff
@@ -28,6 +29,7 @@
     "JaroAtThresholds",
     "JaroWinklerAtThresholds",
     "DistanceFunctionAtThresholds",
+    "PairwiseStringDistanceFunctionAtThresholds",
```
Comment (RobinL) [inline splink/comparison_library.py:32]: Needs the same in splink/comparison_level_library.py

File: tests/test_comparison_lib.py [hunk 2]
```diff
@@ -65,6 +65,60 @@ def test_distance_function_comparison():
             assert sum(df_pred[f"gamma_{col}"] == gamma_val) == expected_count
 
 
+@mark_with_dialects_excluding("sqlite", "spark", "postgres", "athena")
```
[the remark this answers is not in the document]
Comment (zmbc) [reply tests/test_comparison_lib.py:68]: Ah, I see! I saw both kinds of tests but didn't understand the chronology -- that this is a transition in progress. I agree this is much more readable 😄

File: splink/comparison_library.py (hunk 1 above)
Comment (zmbc) [reply splink/comparison_library.py:32]: Right you are!

**Comment by zmbc** [comment]
@RobinL I believe I've addressed your comments. I don't understand why a test is failing -- it does not seem related in any way to these changes.

**Comment by ADBond** [comment]
> @RobinL I believe I've addressed your comments. I don't understand why a test is failing -- it does not seem related in any way to these changes.

@zmbc you are correct - apologies this is an unrelated issue #2515 (which will be fixed shortly, so should not be an issue going forward).

**Pull request merged**